### ISSUE TITLE
Smoke tests: revisit timeouts

### ISSUE THREAD

**Issue by bpasero** [microsoft/vscode#137847]
If we push for less flaky smoke tests, I think we also need to revisit the various `setTimeout` calls we have been adding to smoke tests over the time. A `setTimeout` is a sign that an underlying problem is not well understood and rather than keeping that timeout we should rather find a condition we can use via the driver to assert that a certain state was reached.

Picking some that I found:
* [x] `checkWindowReady` ([source](https://github.com/microsoft/vscode/blob/c28ea165cd1d9d17b0c01322e2a23a7fa4e21561/test/automation/src/application.ts#L143)) @Tyriar @joaomoreno 
* [x] `runCommandInTerminal` ([source](https://github.com/microsoft/vscode/blob/ef440e563d93b76436589864412cd2ac58d492a4/test/automation/src/terminal.ts#L84)) @Tyriar @meganrogge 
* [x] `removeFileMatch` ([source](https://github.com/microsoft/vscode/blob/852b70eafedd9ecf4a6b2587d14845a8ee91c214/test/automation/src/search.ts#L86)) @roblourens 
* [x] `replaceFileMatch` ([source](https://github.com/microsoft/vscode/blob/852b70eafedd9ecf4a6b2587d14845a8ee91c214/test/automation/src/search.ts#L112)) @roblourens 
* [x] `closeQuickInput` ([source](https://github.com/microsoft/vscode/blob/2932ec095dd95631aafb3fbc11aca5731f3ca62c/test/automation/src/quickaccess.ts#L75)) @TylerLeonhardt @chrmarti 
* [x] `closeQuickInput (2)` ([source](https://github.com/microsoft/vscode/blob/2932ec095dd95631aafb3fbc11aca5731f3ca62c/test/automation/src/quickaccess.ts#L118)) @TylerLeonhardt @chrmarti 
* [x] `after` ([source](https://github.com/microsoft/vscode/blob/d7f10626b85fbaaf0b80b20dc4d3887ddf833227/test/smoke/src/main.ts#L338)) @joaomoreno 
* [x] `dispatchKeybinding` ([source](https://github.com/microsoft/vscode/blob/cdfee234401539007f33c87ae0c80b2b5304d381/src/vs/platform/driver/electron-main/driver.ts#L137)) @joaomoreno 
* [x] `click` / `doubleClick` ([source](https://github.com/microsoft/vscode/blob/ee1655a82ebdfd38bf8792088a6602c69f7bbd94/src/vs/platform/driver/electron-sandbox/driver.ts#L35-L38)) @joaomoreno 
* [x] `doubleClick` (web) ([source](https://github.com/microsoft/vscode/blob/8cf4dcdadee0cde35c81b69b94a209d4cf1b3364/test/automation/src/playwrightDriver.ts#L139-L144)) @joaomoreno 
* [x] `backup` ([source](https://github.com/microsoft/vscode/blob/bacf6ed2c1929bba386ddd0286fb9f86a7c3b83b/test/smoke/src/areas/workbench/data-migration.test.ts#L97)) @bpasero

**Comment by Tyriar** [comment]
@bpasero any ideas what we could listen for to make sure the initial "focus steal" has happened to replace this?

https://github.com/microsoft/vscode/blob/c28ea165cd1d9d17b0c01322e2a23a7fa4e21561/test/automation/src/application.ts#L141-L143

**Comment by roblourens** [comment]
The search ones are interesting. They go before a click, and when I remove them, I think I see the issue that we see in https://github.com/microsoft/vscode/issues/137195 - the button appears hovered but the click hasn't taken effect. Like playwright can't click a button unless it waits a bit. I checked that the click handler is added by the time the button is in the DOM

**Comment by roblourens** [comment]
Which process does playwrightDriver.ts run in? If I add a console.log I can't find it anywhere.

**Comment by bpasero** [comment]
@Tyriar please connect with @joaomoreno who I think added this ([commit](https://github.com/microsoft/vscode/commit/2098851c25ffb7f8d2b7c862f3a093b365893a43)). Maybe back in the days we were still opening the welcome editor or release notes? We no longer do that, every smoke test opens with an empty editor area, by configuring some settings from the workspace: https://github.com/microsoft/vscode-smoketest-express/blob/main/.vscode/settings.json

@roblourens please connect with @joaomoreno who might know more about why it was added in https://github.com/microsoft/vscode/commit/bb3aa3770fc5f72081f2027a70d906b32744b39d

**Comment by joaomoreno** [comment]
> @Tyriar please connect with @joaomoreno who I think added this ([commit](https://github.com/microsoft/vscode/commit/2098851c25ffb7f8d2b7c862f3a093b365893a43)). Maybe back in the days we were still opening the welcome editor or release notes? We no longer do that, every smoke test opens with an empty editor area, by configuring some settings from the workspace: https://github.com/microsoft/vscode-smoketest-express/blob/main/.vscode/settings.json

You'd be surprised how many UI pieces steal focus, even if all we do is open an empty editor area. There's a lot of `.focus()` calls in our code base. That timeout was introduced to give the workbench 1 second of _calm down_ time, before starting to click around. This is even something I see in real usage: sometimes after a brand new window I'm really fast at opening quick open only to have it disappear under my fingertips just because some other UI piece called focus.

**Comment by joaomoreno** [comment]
> The search ones are interesting. They go before a click, and when I remove them, I think I see the issue that we see in #137195 - the button appears hovered but the click hasn't taken effect. Like playwright can't click a button unless it waits a bit. I checked that the click handler is added by the time the button is in the DOM

The problem with clicks is that they are not atomic: [they are always a pair of `mousedown` and a `mouseup` events, on the same coordinate](https://github.com/microsoft/vscode/blob/732e263ee9223f015ffbde178c2c2a626c0c5266/src/vs/platform/driver/electron-sandbox/driver.ts/#L31-L40). But the DOM can always change between both events. The `mousedown` can go on the action, but the `mouseup` outside of it. That would explain [why the action appears hovered](https://github.com/microsoft/vscode/issues/137195#issuecomment-975902255). It also explains why waiting helps: it's waiting for the action bar to finish adding all the actions. It also aligns with the fact that the [`ReplaceAllAction` is inserted into the DOM first, and `RemoveAction` comes second](https://github.com/microsoft/vscode/blob/732e263ee9223f015ffbde178c2c2a626c0c5266/src/vs/workbench/contrib/search/browser/searchResultsView.ts/#L197-L200). Finally, it also doesn't help that [actions are right-aligned](https://github.com/microsoft/vscode/blob/732e263ee9223f015ffbde178c2c2a626c0c5266/src/vs/workbench/contrib/search/browser/media/searchview.css/#L233), which means that if the `mousedown` event occurs between both actions go into the DOM, it will hit the `ReplaceAllAction` but the `mouseup` event will hit the `RemoveAction` instead.

In other words: `// ¯\_(ツ)_/¯` because at the time I had had enough of UI automation issues like these. I would appreciate fresh ideas on tackling stuff like this.

**Comment by Tyriar** [comment]
> This is even something I see in real usage: sometimes after a brand new window I'm really fast at opening quick open only to have it disappear under my fingertips just because some other UI piece called focus.

I also see this is real usage, I open the terminal quickly and I think the editor sometimes steals it which is why I asked @bpasero. Once we identify what's causing the focus, maybe we should add a `ready` class or something to a root workbench element so the state can be polled for.

**Comment by Tyriar** [comment]
I can't reproduce any problems locally now when opening an empty workspace, we might be able to just remove that check and assume `body` is focused at the start.

**Comment by roblourens** [comment]
Thanks @joaomoreno. Actually what I think is happening is that the search tree is rerendered multiple times quickly. We find the element once then try to click it as it's being rerendered.

**Comment by bpasero** [comment]
Pushed a change to remove the 500ms timeout:

https://github.com/microsoft/vscode/commit/9d427e4df4428099f568b6dd4300808a271a0a77#diff-aeee5a24db3c5fe340dfff2213f8dfd613393792f78aa67d14c2c02144d5ed29L339

Now that the `exitApplication` waits for the process to be gone (local: Electron main process, remote/web: Server process), I would hope this artificial delay is not needed anymore 🤞

**Comment by Tyriar** [comment]
Trying to remove the `checkWindowReady` timeout in https://github.com/microsoft/vscode/pull/138516, will see if it passes.

**Comment by Tyriar** [comment]
The change to remove `checkWindowReady` in https://github.com/microsoft/vscode/pull/138516 seemed to make the flaky tests even more flaky, so it seems we are not waiting long enough sometimes.

**Comment by Tyriar** [comment]
One of the failures when `checkWindowReady` timeout is removed shows the search happening before the remote connection is finished:
<img width="892" alt="Screen Shot 2021-12-10 at 11 09 44 am" src="https://user-images.githubusercontent.com/2193314/145628602-d8ccacf3-7f55-4c94-a35b-57ba324ed67d.png">

Looks like the remote connection check doesn't apply in web:

https://github.com/Microsoft/vscode/blob/27e38a8e2714b061c1e1647c553ab017f41b99a4/test/automation/src/application.ts#L115-L117

So the timeout was sometimes allowing that to succeed.

**Comment by roblourens** [comment]
There is another issue for that search issue - starting a search before the resolver has finished should be fine and normally works. I've never been able to repro it manually.

**Comment by roblourens** [comment]
Pushed https://github.com/microsoft/vscode/commit/b823936907f86ded04a152a4af202d0fc8423522 to wait for the window to connect in Web tests as well. Starting the search before the window connects should work, but I can't figure out why it isn't, and I think this will reduce the flakiness.

**Comment by bpasero** [comment]
The `doubleClick` implementation for browser really doesn't work when I try it:

https://github.com/microsoft/vscode/blob/8cf4dcdadee0cde35c81b69b94a209d4cf1b3364/test/automation/src/playwrightDriver.ts#L139-L144

I think we are just lucky because the only user is `Explorer.openFile` which is not being used in web so far:

https://github.com/microsoft/vscode/blob/84c895e700f2b3f31a818dbb25de5b53f7866d56/test/automation/src/explorer.ts#L28

I thought I can rewrite that method to:
* single click
* dispatch keybinding `Enter` / `Cmd+Down` (macOS)

But this fails when browser smoke tests run headless 🤯

**Comment by bpasero** [comment]
I was able to remove the double click usage as it was only used from the explorer and no one was calling that `openFile` method.

**Comment by Tyriar** [comment]
I removed the `checkWindowReady` timeout as there is now an `isWeb` case and the PR passes CI https://github.com/microsoft/vscode/pull/138516. If this ends up causing more flakes that will indicate we're not waiting for some other event.
